### PR TITLE
fix: support dynamic resource key names

### PR DESCRIPTION
**Pull request by shemau**
### Description

#601 support for dynamic resource key names

### Release required?

- [ ] No release
- [x] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

A new optional `key_name` has been added to `var.resource_keys` objects. When specified this will be used as the resource key name and can be a dynamic reference.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by shemau** [comment]
The upgrade test fails because the test changes the static reference name of the key to a dynamic reference to demonstrate the new feature. Skipping upgrade.

This is the line:

```hcl
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ name                  = "cos-upgrade-crujbk-writer-key" -> "iam-ServiceId-9bc62260-22db-4c31-834d-c137673f1928" # forces replacement
```

From:

```hcl
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185: Terraform will perform the following actions:
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185: 
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:   # module.cos_bucket1.ibm_resource_key.resource_keys["cos-upgrade-crujbk-writer-key"] must be replaced
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185: -/+ resource "ibm_resource_key" "resource_keys" {
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ account_id            = "abac0df06b644a9cabc6e44f55b3880e" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ created_at            = "2024-05-04T20:32:05.006Z" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ created_by            = "IBMid-666000KAO3" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ credentials           = (sensitive value)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ credentials_json      = (sensitive value)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ crn                   = "crn:v1:bluemix:public:cloud-object-storage:global:a/abac0df06b644a9cabc6e44f55b3880e:3da653a6-355c-42a3-ae97-941fbf4919f5:resource-key:220a4dce-b780-47da-ac9f-67334856ef9b" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       + deleted_at            = (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       + deleted_by            = (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ guid                  = "220a4dce-b780-47da-ac9f-67334856ef9b" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ iam_compatible        = true -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ id                    = "crn:v1:bluemix:public:cloud-object-storage:global:a/abac0df06b644a9cabc6e44f55b3880e:3da653a6-355c-42a3-ae97-941fbf4919f5:resource-key:220a4dce-b780-47da-ac9f-67334856ef9b" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ name                  = "cos-upgrade-crujbk-writer-key" -> "iam-ServiceId-9bc62260-22db-4c31-834d-c137673f1928" # forces replacement
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ resource_group_id     = "644e45febd4d45168e6f557e8d661caf" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ resource_instance_url = "/v2/resource_instances/3da653a6-355c-42a3-ae97-941fbf4919f5" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ source_crn            = "crn:v1:bluemix:public:cloud-object-storage:global:a/abac0df06b644a9cabc6e44f55b3880e:3da653a6-355c-42a3-ae97-941fbf4919f5::" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ state                 = "active" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ status                = "active" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ updated_at            = "2024-05-04T20:32:05.006Z" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       + updated_by            = (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:       ~ url                   = "/v2/resource_keys/220a4dce-b780-47da-ac9f-67334856ef9b" -> (known after apply)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:         # (3 unchanged attributes hidden)
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185:     }
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185: 
TestRunUpgradeExample 2024-05-04T20:33:36Z command.go:185: Plan: 1 to add, 0 to change, 1 to destroy.
```

File: variables.tf [hunk 1]
```diff
@@ -18,10 +18,16 @@ variable "create_cos_instance" {
   default     = true
 }
 
+# 'name' is the terraform static reference to variablthe objecy in the list
```
Comment (vburckhardt) [inline variables.tf:21]: small typo here

**Review by toddgiguere — APPROVED** [review]

**Pull request merged**